### DISCUSSION TITLE
rfsrc(Surv(...) does NOT work on my own dataset

### DISCUSSION THREAD

**Discussion by Marina300** [kogalur/randomForestSRC#123] [Q&A]
I am using the R Package ‘randomForestSRC’ 

I wish to use my own dataset ([yellowCOPD27patients.csv](https://github.com/kogalur/randomForestSRC/files/7562625/yellowCOPD27patients.csv)) for the following instruction
**marina.obj <- rfsrc(Surv(To, Exacerbation_Event) ~ ., yellow_df)**

I get the following error;
_Error in finalizeData(c(subj.names, yvar.names, xvar.names), data, na.action,  :
  data types cannot be character: please convert all characters to factors_

All the data in the dataset are "integer" and not "character" and the conversion into factor using "as.factor" does not work. 

I am sure rfsrc requires a type of object that in different from what I am providing. 
However, when I use Surv() and coxph I do not get any problem at all:
**surv_object<-Surv(time = yellow_df$To, event=yellow_df$Exacerbation_Event)
cox1 <- coxph(surv_object ~ age, data = yellow_df)**

Additionally, when I specify
**rfsrc(surv_object ~ age, data = yellow_df)**
I get the following error:
_Error in parseFormula(formula, data, ytry) :
  formula is incorrectly specified.__

Would you be so kind to help me with this problem?

Thanks in advance. 

Best Regards
Marina

**Comment by ishwaran** [comment]
I believe both errors are related to the same problem that you are not correctly specifying the model.  As described in the manual, and also see "example grow calls" on our webpage (https://luminwin.github.io/randomForestSRC/articles/getstarted.html), a random survival analysis is initiated by a call of the form:

`rfsrc(Surv(To, Exacerbation_Event) ~ ., yellow_df)`

**Comment by Marina300** [comment]
Thanks a lot for your answer! I think I understood where the error lies: when one of the columns of the dataset has a decimal value (eg. 57.8),  the following error is returned: 
_Error in finalizeData(c(subj.names, yvar.names, xvar.names), data, na.action,  :
  data types cannot be character: please convert all characters to factors_

When I remove the decimals and use only integers, the error disappears. 
In the file that I attached there are no decimals. But if you add a column with decimals, you will be able to replicate the error. 

The file that I attached was a excerpt from a larger file. In the larger file one of the columns contained decimals. 

Cheers,
Marina

**Comment by ishwaran** [comment]
The package handles real values for the X features.  Having a decimal value in the data values for a covariate should not be causing the error.

I believe your error has to do with the response (survival time and censoring) but more specifically the way you are specifying the formula which appears incorrect.

In any case, please provide a reproducible example here online that can be cut and pasted.  Otherwise there is no way to proceed further.

**Comment by Marina300** [comment]
Hi, 
in which way should I specify the formula? The dataset is attached to my initial question and is called: yellowCOPD27patients.csv
I run this code and it works: 
_rm(list = ls(all.names = TRUE)) #will clear all objects includes hidden objects.
yellow_df <- read.csv("yellowCOPD27patients.csv", 
                       header = TRUE, sep = ";", na.strings = "NA", dec = ".", 
                       strip.white = TRUE,fileEncoding="UTF-8-BOM")
colnames(yellow_df)
dim(yellow_df)
rfsrc(Surv(To, Exacerbation_Event) ~ ., yellow_df)_

I get the following output: 
_Sample size: 6571
                    Number of deaths: 104
                     Number of trees: 500
           Forest terminal node size: 15
       Average no. of terminal nodes: 121.218
No. of variables tried at each split: 3
              Total no. of variables: 7
       Resampling used to grow trees: swor
    Resample size used to grow trees: 4153
                            Analysis: RSF
                              Family: surv
                      Splitting rule: logrank *random*
       Number of random split points: 10
                    (OOB) Error rate: 14.59609944%_

When I add the column "weight" that contains decimals to yellowCOPD27patients.csv (this is the dataset yellow+weight:
[yellow_weight_COPD27patients.csv](https://github.com/kogalur/randomForestSRC/files/7571510/yellow_weight_COPD27patients.csv), and I run the same code as above, ie

__rm(list = ls(all.names = TRUE)) #will clear all objects includes hidden objects.

yellow_weight_df <- read.csv("yellow_weight_COPD27patients.csv", 
                      header = TRUE, sep = ";", na.strings = "NA", dec = ".", 
                      strip.white = TRUE,fileEncoding="UTF-8-BOM")

colnames(yellow_weight_df)
dim(yellow_weight_df)

rfsrc(Surv(To, Exacerbation_Event) ~ ., yellow_weight_df)__

I get the following error: 
**Error in finalizeData(c(subj.names, yvar.names, xvar.names), data, na.action,  : 
  data types cannot be character: please convert all characters to factors**

Thank you for pointing out that package can handle decimal values, because I realized that when reading the dataset I specified the dot as decimal separator, but I actually have a comma (Swedish decimal separator). The cvs has been saved with the Swedish locale. 

I fixed this and now it works smoothly. 

If you think I am specifying the model in the wrong way, pls let me know the correct syntax. Is the output that I get with theyellowCOPD27patients.csv dataset reliable or not?

Please do not hesitate to ask more information, if I have been unclear in formulating my problem. 

Thanks in advance for your answer. 

Cheers, Marina

**Comment by ishwaran** [comment]
OK, the output looks sensible so it looks like it's working now.  Thus it was an issue with the formatting of your data due to use of "," instead of decimal ".".

**Reply by Marina300** [reply]
Yes :-)
